Model/ResourceModel/Product/Index/Collection.php

```diff
@@ -0,0 +1,329 @@
+            ['eq' => Index::DB_VALUE_BOOLEAN_TRUE]
+        );
+    }
+
+    /**
+     * Returns the first item of the collection
+     * or null if the collection is empty
+     *
+     * @return ProductIndexInterface|null
+     * @suppress PhanTypeMismatchReturn
+     */
+    public function getOneOrNull()
+    {
+        $this->getSelect()->limit(1);
+        return $this->getSize() > 0 ? $this->getLastItem() : null; // @codingStandardsIgnoreLine
+    }
+
+    /**
+     * Marks products as deleted by given product ids and store
+     *
+     * @param array $ids
+     * @param Store $store
+     * @return int
+     */
+    public function markAsDeleted(array $ids, Store $store)
+    {
+        if (empty($ids)) {
+            return 0;
+        }
+        $connection = $this->getConnection();
+        return $connection->update(
+            $this->getMainTable(),
+            [Index::IS_DELETED => Index::DB_VALUE_BOOLEAN_TRUE],
+            [
+                sprintf('%s IN (?)', Index::PRODUCT_ID) => array_unique($ids),
+                sprintf('%s=?', Index::STORE_ID) => $store->getId()
+            ]
+        );
+    }
+
+    /**
+     * Marks products as deleted by given product ids and store
+     *
+     * @param array $productIds array of product ids
+     * @param Store $store
+     * @return int
+     */
+    public function markAsInSync(array $productIds, Store $store)
+    {
+        if (empty($productIds)) {
+            return 0;
+        }
+        $connection = $this->getConnection();
+        return $connection->update(
+            $this->getMainTable(),
+            [Index::IN_SYNC => Index::DB_VALUE_BOOLEAN_TRUE],
+            [
+                sprintf('%s IN (?)', Index::PRODUCT_ID) => array_unique($productIds),
+                sprintf('%s=?', Index::STORE_ID) => $store->getId()
+            ]
+        );
+    }
+
+    /**
+     * Marks all products as dirty by given Store
+     *
+     * @param Store $store
+     * @return int
+     */
+    public function markAllAsDirtyByStore(Store $store)
+    {
+        $connection = $this->getConnection();
+        return $connection->update(
+            $this->getMainTable(),
+            [Index::IS_DIRTY => Index::DB_VALUE_BOOLEAN_TRUE],
+            [
+                sprintf('%s=?', Index::STORE_ID) => $store->getId()
+            ]
+        );
+    }
+
+    /**
+     * Deletes current indexed products in store
+     *
+     * @param Store $store
+     * @return int
+     */
+    public function deleteCurrentItemsByStore(Store $store)
+    {
+        if ($this->getSize() === 0) {
+            return 0;
+        }
+        $indexIds = [];
+        /* @var Index $item */
+        foreach ($this->getItems() as $item) {
+            $indexIds[] = $item->getId();
+        }
+        $connection = $this->getConnection();
+        return $connection->delete(
+            $this->getMainTable(),
+            [
+                sprintf('%s IN (?)', Index::ID) => array_unique($indexIds),
```

Comment:
Is this really the recommended way of buillding queries?

Reply:
Let's check if we should do this one by one or if there's a cleaner way to do these mass operations